docs/gni_sources.md

```diff
@@ -1,3 +1,9 @@
+### sources.gni
 
-This does not mean that you cannot use sources.gni to inject sources into existing targets, but please try to avoid it when practical.
+Use of sources.gni to include sources in upstream targets should should generally be avoided. Whenever possible try to break circular dependencies see [Recipes for Breaking Chrome Dependencies](https://www.chromium.org/developers/design-documents/cookbook/#recipes-for-breaking-chrome-dependencies) and [https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion](https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion) for examples. An interface/implementation pattern can also often be used where header files and possibly some cc files are included in the direct dependency and the code that would cause the circular dependency is included in a higher level target like `//brave/browser` to ensure that the implementation code is always linked into the final output. See [tabs:tabs_public](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=12) and [tabs:impl](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=304) and [//chrome/browser impl dependency](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=4354). The chromium ios code is also a good model for separating out dependencies and sometimes makes use of interface/implementation patterns.
+
+Circular dependencies can sometimes (temporarily) use `allow_circular_includes_from` to split sources up into smaller targets so they can be more easily resolved down the road. See [`//chrome/browser/ui`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=3524) and [`//chrome/browser/ui`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/BUILD.gn;l=5755) for examples.
+
+This does not mean that you cannot ever use sources.gni. For instance it may be appropriate when adding a very small number of sources to an existing upstream target, but please consider other approaches first. Using sources.gni to add dependencies and other non-source configuration to upstream targets is also ok.
+
+Do not use `check_includes = false`.
```

Comment:
iOS uses `check_includes` in **one** place: https://github.com/brave/brave-core/blob/master/ios/browser/svg/BUILD.gn#L16 because Skia in Chromium does the same: https://source.chromium.org/chromium/chromium/src/+/main:skia/BUILD.gn;l=671

Reply:
we can make one-off exceptions in rare cases